### PR TITLE
Context Parallel SFT Support for dataset in THD format

### PR DESCRIPTION
**Pull request by tomlifu**
# What does this PR do ?

This PR adds CP support for THD format and is compatible with cu_seqlen_padded in the latest CUDNN fused attention.

Steps to run SFT + CP + THD format:
1. Prepare packed dataset in THD format: run `scripts/nlp_language_modeling/prepare_packed_ft_dataset.py` to pack the dataset into THD format in desired sequence length. For example:
```
python <NeMo_top_dir>/scripts/nlp_language_modeling/prepare_packed_ft_dataset.py \
        model.data.train_ds.file_names=[<dataset_top_dir>/squad/1_squad_train.jsonl] \
        model.data.train_ds.max_seq_length=4096 \
        +model.context_parallel_size=2 \
        +tokenizer_path=<tokenizer_path> \
        +output_dir=<output_dir> +pack_sizes=[4096] \
```
2. Run SFT on the packed dataset in THD format with the same CP size specified in the last step

**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

### CODE REVIEW

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py [hunk 1]
```diff
@@ -618,17 +618,27 @@ def collate_fn(self, batch):
...
                 position_ids[-1].extend(list(range(l - 1)))
                 cu_seqlens[-1].append(cu_seqlens[-1][-1] + l - 1)
+
             # set last seq to the max seq len because rope and attn kernels expect no padding
             cu_seqlens[-1][-1] = max_length
 
+            for i in range(len(item['seq_boundaries']) - 1):
+                seqlen_unpadded = np.argmin(
+                    np.array(item['input_ids'][item['seq_boundaries'][i] : item['seq_boundaries'][i + 1] - 1])
+                )
```
Comment (xrennvidia) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:638]: We assume pad_id is the minimum value in `input_ids`? is this always true?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 2]
```diff
@@ -1303,13 +1310,34 @@ def fwd_output_and_loss_func(dataloader_iter, model, checkpoint_activations_all_
...
-                    )
+                    # get packed sequences for this context parallel rank
+                    cp_size = parallel_state.get_context_parallel_world_size()
+                    if cp_size > 1:
+                        cp_rank = parallel_state.get_context_parallel_rank()
+                        for key in required_keys:
+                            val = batch[key]
+                            if key not in {"cu_seqlens", "cu_seqlens_unpadded"}:
+                                index = tex.thd_get_partitioned_indices(cu_seqlens, val.size(1), cp_size, cp_rank)
+                                val = val.index_select(1, index)
```
Comment (xrennvidia) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:1321]: Why 1, not 0? val shape should be [t, h, d], and we should index in dimension of t?

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py (hunk 1 above)
Comment (tomlifu) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:643]: I used the dataset.tokenizer.eos_id as pad_id and it's the minimum value in input_ids.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py (hunk 2 above)
Comment (tomlifu) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:1322]: The THD dataset is in [1, t, h, d] in this case. Thus we use 1 here.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 3]
```diff
@@ -1303,13 +1310,34 @@ def fwd_output_and_loss_func(dataloader_iter, model, checkpoint_activations_all_
...
+                                batch[key] = val
+                        cu_seqlens = cu_seqlens // cp_size
+                        cu_seqlens_unpadded = cu_seqlens_unpadded // cp_size
```
[the remark this answers is not in the document]
Comment (tomlifu) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py]: I've pushed a new commit addressing this issue.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py [hunk 4]
```diff
@@ -618,17 +618,27 @@ def collate_fn(self, batch):
...
+                cu_seqlens_unpadded[-1].append(cu_seqlens_unpadded[-1][-1] + seqlen_unpadded)
+            cu_seqlens_unpadded[-1][-1] = max_length
```
[the remark this answers is not in the document]
Comment (tomlifu) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py]: I've made a new commit addressing this issue.

File: scripts/nlp_language_modeling/prepare_packed_ft_dataset.py [hunk 5]
```diff
@@ -121,7 +129,31 @@ def tokenize_dataset(cfg: 'DictConfig'):
...
+        def pre_pad_dataset(data, max_length, pad_id):
+            '''
+            pad each individual data point to the length of max_length
+            '''
+            for key, val in data.items():
+                if key in {'input_ids', 'context_ids'}:
+                    # because input_ids is truncated by 1 for labels in the collate_fn of GPTSFTPackedDataset
+                    # in gpt_sft_dataset.py, we add 1 extra padding here
+                    val = val + [pad_id] * (max_length - len(val) + 1)
+                    data[key] = val
```
[the remark this answers is not in the document]
Comment (tomlifu) [reply scripts/nlp_language_modeling/prepare_packed_ft_dataset.py]: I've added changes that supports truncation in a new commit.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 6]
```diff
@@ -1303,13 +1310,34 @@ def fwd_output_and_loss_func(dataloader_iter, model, checkpoint_activations_all_
...
+                        }
+                        forward_args['packed_seq_params'] = PackedSeqParams(
+                            cu_seqlens_q=cu_seqlens_unpadded,
+                            cu_seqlens_kv=cu_seqlens_unpadded,
+                            cu_seqlens_q_padded=cu_seqlens,
+                            cu_seqlens_kv_padded=cu_seqlens,
+                            max_seqlen_q=max_seqlen,
+                            max_seqlen_kv=max_seqlen,
+                            qkv_format='thd',
+                        )
```
[the remark this answers is not in the document]
Comment (tomlifu) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py]: Fixed in the latest commit.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py [hunk 7]
```diff
@@ -618,17 +618,28 @@ def collate_fn(self, batch):
...
+            cu_seqlens_unpadded.append([0])
             seqlens = np.array(item['seq_boundaries'][1:]) - np.array(item['seq_boundaries'][:-1])
             for l in seqlens:
                 # length minus 1 because input_ids is truncated by 1 for labels
                 position_ids[-1].extend(list(range(l - 1)))
                 cu_seqlens[-1].append(cu_seqlens[-1][-1] + l - 1)
-            # set last seq to the max seq len because rope and attn kernels expect no padding
+
+            # the last seq needs to be the max seq len because rope and attn kernels expect no padding
             cu_seqlens[-1][-1] = max_length
```
Comment (xrennvidia) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:633]: Noticed that you have `cu_seqlens_unpadded`, so your `cu_seqlens` is actually `cu_seqlen_padded`? If so, I have same question as before, why overwrite the last element directly? shouldn't you append max_length instread?

I see you did this in [this commit](https://github.com/NVIDIA/NeMo/pull/10688/commits/850d9ae5ac485d8a1049981e67ccd008e5622a29), why did you recover the change later?

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py [hunk 8]
```diff
@@ -618,17 +618,28 @@ def collate_fn(self, batch):
...
 
+            for i in range(len(item['seq_boundaries']) - 1):
+                seqlen_unpadded = np.where(
+                    np.array(item['input_ids'][item['seq_boundaries'][i] : item['seq_boundaries'][i + 1] - 1])
+                    == self.tokenizer.eos_id
+                )
+                seqlen_unpadded = seqlen_unpadded[0][1] if seqlen_unpadded[0].size > 1 else seqlen_unpadded[0][0]
```
Comment (xrennvidia) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:640]: what does this line of code mean? why not always `seqlen_unpadded[0][0]`?

File: scripts/nlp_language_modeling/prepare_packed_ft_dataset.py [hunk 9]
```diff
@@ -121,7 +129,42 @@ def tokenize_dataset(cfg: 'DictConfig'):
...
+    if cp_size > 1:
+
+        def pre_pad_dataset(data, max_seq_length, max_length_to_pad, pad_id):
+            '''
+            pad each individual data point to the length of max_length
+            '''
+            assert max_seq_length >= max_length_to_pad
+            for key, val in data.items():
+                if key in {'input_ids', 'context_ids'}:
+                    if len(val) < max_seq_length:
```
Comment (xrennvidia) [inline scripts/nlp_language_modeling/prepare_packed_ft_dataset.py:145]: should be `if len(val) < max_length_to_pad`?

File: scripts/nlp_language_modeling/prepare_packed_ft_dataset.py [hunk 10]
```diff
@@ -121,7 +129,42 @@ def tokenize_dataset(cfg: 'DictConfig'):
...
+                        # because input_ids are truncated by 1 for inputs and labels,
+                        # we add 1 extra padding here to make sure padded inputs and labels
+                        # are is a multiple of (cp_size * 2)
+                        val = val + [pad_id] * (max_length_to_pad - len(val) + 1)
+                        data[key] = val
+                    else:
```
Comment (xrennvidia) [inline scripts/nlp_language_modeling/prepare_packed_ft_dataset.py:151]: should be `elif len(val) > max_seq_legnth`?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 11]
```diff
@@ -24,11 +24,13 @@
 
 import packaging
 import torch
+import transformer_engine_torch as tex
 from omegaconf import OmegaConf
 from omegaconf.dictconfig import DictConfig
 from pytorch_lightning.accelerators import CPUAccelerator
 from pytorch_lightning.loops.fetchers import _DataFetcherWrapper
 from pytorch_lightning.trainer.trainer import Trainer
+from transformer_engine.pytorch.attention import get_cu_seqlens_on_cp_rank
```
[the remark this answers is not in the document]
Comment (tomlifu) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py]: addressed in the latest commit

File: scripts/nlp_language_modeling/prepare_packed_ft_dataset.py [hunk 12]
```diff
@@ -121,7 +129,42 @@ def tokenize_dataset(cfg: 'DictConfig'):
...
+                                        larger than the max_seq_length specified ({max_seq_length}).
+                                        The current seqquence length is truncated to the size of max_seq_length.
+                                        Please consider increase the sequence packing size"""
+                        )
+                        data[key] = val[:max_seq_length]
+            return
+
+        ceil_to_nearest = lambda n, m: (n + m - 1) // m * m
+        for data in dataset:
+            max_length = min(max_seq_length, ceil_to_nearest(len(data['input_ids']), pad_seq_length_to_mult))
```
[the remark this answers is not in the document]
Comment (tomlifu) [reply scripts/nlp_language_modeling/prepare_packed_ft_dataset.py]: changed max_length to max_length_to_pad in the latest commit

File: scripts/nlp_language_modeling/prepare_packed_ft_dataset.py (hunk 9 above)
Comment (tomlifu) [reply scripts/nlp_language_modeling/prepare_packed_ft_dataset.py]: Yes, I've updated it in the latest commit.

File: scripts/nlp_language_modeling/prepare_packed_ft_dataset.py (hunk 10 above)
Comment (tomlifu) [reply scripts/nlp_language_modeling/prepare_packed_ft_dataset.py]: addressed in the latest commit

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py (hunk 8 above)
Comment (tomlifu) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:640]: The padding added in the data preparation code is EOS token. We need to make sure that one EOS token is not counted as padding. Thus it's not always `seqlen_unpadded[0][0]`.

Comment (xrennvidia) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:640]: I think I do not know what is `seqlen_unpadded[0].size`? why do we need to check if it's bigger than 1? could you please elaborate this? thanks

Comment (tomlifu) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py]: I have pushed a new commit renaming the some variables and adding some comments. Hopefully it will be more clear.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py (hunk 7 above)
Comment (tomlifu) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py]: Not sure why it was left out. I have reverted this part back to what I had in the [this commit](https://github.com/NVIDIA/NeMo/pull/10688/commits/850d9ae5ac485d8a1049981e67ccd008e5622a29)

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py (hunk 8 above)
Comment (xrennvidia) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py]: Thanks, I am just thinking all tokens are padded at the end, so if `eos_idx[0].size > 1`, isn't `eos_idx[0][1] == eos_idx[0][0] + 1`? If yes, shouldn't we just do `seqlen_unpadded = eos_idx[0][0] + 1`?

**Comment by xrennvidia** [comment]
Please fix [DCO](https://github.com/NVIDIA/NeMo/pull/10688/checks?check_run_id=31972432771) also.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py (hunk 8 above)
Comment (tomlifu) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py]: Yes, I just updated that.

**Comment by switiz** [comment]
Could you let me know when it will be completed? I’ve been really looking forward to this feature. It works in pretrain, but it’s really strange that it doesn’t work in SFT.

File: nemo/utils/sequence_packing_utils.py [hunk 13]
```diff
@@ -115,7 +115,7 @@ def create_hist(dataset: np.array, truncate_seq_len: int):
     logging.info("Creating histogram from tokenized dataset...")
 
     sequences = collections.defaultdict(list)
-    counts = [0] * truncate_seq_len
+    counts = [0] * (truncate_seq_len + 1)
```
Comment (cuichenx) [inline nemo/utils/sequence_packing_utils.py:118]: thanks for fixing this, line 132 below should need the same fix

File: scripts/nlp_language_modeling/prepare_packed_ft_dataset.py [hunk 14]
```diff
@@ -88,6 +88,14 @@ def tokenize_dataset(cfg: 'DictConfig'):
     # using the same template as SFT/PEFT script. This may be overkill but guarantees the preprocess settings
     # are identical to normal SFT training
     data_cfg = cfg.model.data.train_ds
+    pad_seq_length_to_mult = 16
+    cp_size = cfg.model.context_parallel_size
```
[the remark this answers is not in the document]
Comment (tomlifu) [reply scripts/nlp_language_modeling/prepare_packed_ft_dataset.py]: Thanks Chen. I've added the guard in the latest commit

File: nemo/utils/sequence_packing_utils.py (hunk 13 above)
Comment (tomlifu) [reply nemo/utils/sequence_packing_utils.py:118]: Done.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py [hunk 15]
```diff
@@ -622,16 +630,40 @@ def collate_fn(self, batch):
...
+
+                # The second eos_id index marks the length of the original unpadded sequence if the sequence is
+                # prepadded for cp_size > 1. Otherwise, there is no extra padding.
+                seqlen_unpadded = eos_idx[0][0] + 1 if eos_idx[0].any() else len(current_seq)
+                cu_seqlens_unpadded[-1].append(cu_seqlens_unpadded[-1][-1] + seqlen_unpadded)
+
+            # if extra paddings are added in the packed sequence, they can't be counted as
+            # actual tokens for training
+            if len(cu_seqlens[-1]) > len(cu_seqlens_unpadded[-1]):
+                cu_seqlens_unpadded[-1].append(cu_seqlens_unpadded[-1][-1])
```
Comment (xrennvidia) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:666]: What does this do? why we need this?
Will this make `len(cu_seqlens) == len(cu_seqlens_unpadded)` always True? and then [this check](https://github.com/NVIDIA/NeMo/pull/10688/files#diff-e93ccae74f4b67d341676afc9f3c7e2c50f751ec64df84eb3b2a86b62029ef76R1356) does not work any more?

Comment (tomlifu) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:666]: We need this because the last seq (`cu_seqlens[-1][-1]` ) needs to be the max seq len for RoPE. Therefore I need to find a way to tell the attention to exclude the last bit of padding for attention calculation. And you are right. This will make  len(cu_seqlens) == len(cu_seqlens_unpadded), and we don't need [this check](https://github.com/NVIDIA/NeMo/pull/10688/files#diff-e93ccae74f4b67d341676afc9f3c7e2c50f751ec64df84eb3b2a86b62029ef76R1356) anymore.

Comment (xrennvidia) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:666]: synced offline.

**Review by xrennvidia — APPROVED** [review]
LGTM. Thanks.
